### ISSUE TITLE
ukify should copy MajorImageVersion from vmlinuz into uki

### ISSUE THREAD

**Issue by xnox** [systemd/systemd#30086]
### Is your feature request related to a problem? Please describe

when inspecting kernels which look like pe files (either uki or vmlinuz with efi stub) lots of tools look at PE optional header settings

```
MajorImageVersion       3
MinorImageVersion       0
```

to determine if the vmlinuz supports certain features or not.

However it seems like UKIs have that set to zero

### Describe the solution you'd like

Would it be possible to inspect vmlinuz, and copy its MajorImageVersion & MinorImageVersion into the assembled uki?

this way inspecting uki will expose the features underlying vmlinuz supports.

this would be useful for example for grub to launch uki as if it was vmlinuz via all the latest startinmage() efi APIs.

**Comment by ardbiesheuvel** [comment]
There are currently 2 features that are exposed via the image version field:
- whether the image supports initrd loading via LF2
- whether the image is guaranteed to support initrd loading via the command line with the extended semantics (refer to linux commit e346bebbd36b1576a3335331fed61bb48c6d8823 "efi: libstub: Always enable initrd command line loader and bump version")

Are any of these relevant for UKI?

**Comment by xnox** [comment]
@ardbiesheuvel yes they are - because that changes how grub launches a UKI when it is specified as linux command or as chainload command. And it thus changes measurements.

Separate it may enable distributions signing initrd-less UKI and have that working correctly (such that kernel + cmdline + dtb are signed & measured, with initrd only measured) to achieve systems closer to on-host generated initrd whilst using some of the benefits that a fixed UKI provides.

**Comment by ardbiesheuvel** [comment]
Is there any actual difference when GRUB does not provide an initrd?

Perhaps the version should remain cleared for UKIs that implement LoadFile2 themselves. A similar issue came up with iPXE a while ago where a conflict resulted from the fact that both iPXE and UKI attempted to install the LoadFile2 initrd device path (which the UEFI spec does not permit, and would result in ambiguity even if it did)

**Comment by xnox** [comment]
> Is there any actual difference when GRUB does not provide an initrd?
> 

On x86 - it either sets up LoadFile2 or passes in legacy initrd kernel efi stub to the UKI which may or may not pass it to the kernel stub (didn't check that yet).

On arm64 - it either sets up LoadFile2 or passes in initrd via devicetree

I agree both behaviours are undesirable for UKI with an initrd section.
But it is very much desirable to launch UKI with a loadfile2 provided initrd by grub, if and only if UKI does *not* contain initrd section.

> Perhaps the version should remain cleared for UKIs that implement LoadFile2 themselves. A similar issue came up with iPXE a while ago where a conflict resulted from the fact that both iPXE and UKI attempted to install the LoadFile2 initrd device path (which the UEFI spec does not permit, and would result in ambiguity even if it did)

I agree this is bad. I guess the point is we need to be able to differentiate between vmlinux-PE and UKI and all other things. Cause right now, all zeros for ( Major/Minor OSystemVersion, ImageVersion, SubsystemVersion ) could be really old linux; old shim; or grub; or UKI.

I don't know if we can set Major Image Version to like 9 for UKI, and Minor one to an indicator of which sections an image has and what sort of things it implements.

I could parse and iterate sections of the UKI in grub to check if it does or doesn't have .linux to determine it is a UKI, and check if it does or doesn't have .initrd, and provide one, but this seems like fragile and something declarative via ImageVersion would be better.
